mdf_matio/validator.py

```diff
@@ -0,0 +1,397 @@
+                    "details": ("Not authorized for project block(s) '{}' in '{}'. "
+                                "The dataset is not in an allowed organization."
+                                .format(unauthorized, ds_md))
+                }
+
+        # Validate required fields
+        # TODO: How should this validation be done?
+        # The metadata conforms to the schema, there are just extra
+        # `requires` values. Perhaps add these to the schema instead?
+        # Lists, specifically, are an issue. Must all dicts in the list
+        # conform? This behavior is difficult.
+        # As a semi-temporary measure, only check the first element of lists.
+        if self.__required_fields:
+            missing = []
+            for field_path in self.__required_fields:
+                value = ds_md
+                for field_name in field_path.split("."):
+                    try:
+                        value = value[field_name]
+                        if isinstance(value, list) and len(value) > 0:
+                            value = value[0]
+                    except KeyError:
+                        missing.append(field_path)
+                        break
+            if missing:
+                return {
+                    "success": False,
+                    "error": "Missing organization metadata: '{}' are required".format(missing),
+                    "details": ("Required fields are '{}', but '{}' are missing"
+                                .format(self.__required_fields, missing))
+                }
+
+        # Create temporary file for records
+        self.__tempfile = TemporaryFile(mode="w+")
+
+        # Save dataset metadata
+        # Also ensure metadata is JSON-serializable
+        self.__dataset = json.loads(json.dumps(ds_md))
+
+        # Return results
+        return {
+            "success": True
+            }
+
+    def add_record(self, rc_md):
+        """Validate a record against the MDF schema.
+
+        Arguments:
+        rc_md (dict): The record metadata to validate.
+
+        Returns:
+        dict: success (bool): True on success, False on failure
+            If success is False:
+              error (str): A short message about the error.
+              details (str): The full jsonschema error message.
+        """
+        if self.__finished:
+            return {
+                "success": False,
+                "error": ("Dataset has been finished by calling get_finished_dataset(),"
+                          " and no more records may be entered.")
+                }
+        elif not self.__dataset:
+            return {
+                "success": False,
+                "error": "Dataset not started."
+                }
+
+        # Load schema
+        with open(os.path.join(self.__schema_dir, "record.json")) as schema_file:
+            schema = json.load(schema_file)
+        resolver = jsonschema.RefResolver(base_uri="file://{}/".format(self.__schema_dir),
```

Comment:
Make your ref-resolver a module-level global variable or, at least, a class attribute of the validator. They do cache schemas but this the cache dies with this object.

Reply:
This is resolved with the `mdf-validation` merge. The `RefResolver` is now created in the `__init__`. Making it a module-level global would run into the same issues as the `GenericMDFAdapter` above.